Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp

```diff
@@ -2,24 +2,38 @@
 // Licensed under GPLv2+
 // Refer to the license.txt file included.
 
+#include <thread>
+
 #include <QMouseEvent>
 #include <QRegExp>
 #include <QString>
-#include <thread>
 
 #include "DolphinQt2/Config/Mapping/MappingButton.h"
 
 #include "Common/Thread.h"
+#include "DolphinQt2/Config/Mapping/IOWindow.h"
+#include "DolphinQt2/Config/Mapping/MappingCommon.h"
 #include "DolphinQt2/Config/Mapping/MappingWidget.h"
 #include "DolphinQt2/Config/Mapping/MappingWindow.h"
 #include "InputCommon/ControlReference/ControlReference.h"
 #include "InputCommon/ControllerEmu/ControllerEmu.h"
 #include "InputCommon/ControllerInterface/ControllerInterface.h"
+#include "InputCommon/ControllerInterface/Device.h"
+
+static QString TrunicateDisplayString(const QString& string)
+{
+  if (string.length() < 16)
+    return string;
+  else
+    return string.left(13).append(QStringLiteral("..."));
+}
 
 MappingButton::MappingButton(MappingWidget* widget, ControlReference* ref)
-    : QPushButton(QString::fromStdString(ref->expression)), m_parent(widget), m_reference(ref)
+    : QPushButton(TrunicateDisplayString(QString::fromStdString(ref->expression))),
+      m_parent(widget), m_reference(ref),
+      m_io_window(
```

Comment:
Would you mind trying it out? The point of Qt's signals-and-slots things is to make it easier to do this sort of thing. If you think about it in the MVC paradigm, you can think of it like a PushButton is a dumb view, so you don't want it owning other views, and want the controller/model to handle all of the logic.